akd/src/storage/transaction.rs

```diff
@@ -13,25 +13,23 @@ use crate::storage::types::ValueState;
 use crate::storage::types::ValueStateRetrievalFlag;
 use crate::storage::Storable;
 
+use dashmap::DashMap;
 use log::{debug, error, info, trace, warn};
 use std::collections::HashMap;
+use std::sync::Arc;
 use tokio::sync::RwLock;
 
-#[derive(Default, Clone)]
-struct TransactionState {
-    mods: HashMap<Vec<u8>, DbRecord>,
-    active: bool,
-}
-
 /// Represents an in-memory transaction, keeping a mutable state
 /// of the changes. When you "commit" this transaction, you return the
 /// collection of values which need to be written to the storage layer
 /// including all mutations. Rollback simply empties the transaction state.
+#[derive(Clone)]
 pub struct Transaction {
-    state: RwLock<TransactionState>,
+    mods: Arc<DashMap<Vec<u8>, DbRecord>>,
+    active: Arc<RwLock<bool>>,
```

Comment:
Ah maybe the use of a rwlock here is hurting us in parallel mode. Try an atomic bool instead! You could also convert the counters to atomic u64s

Reply:
Ah I gave it a shot, it only improves execution time marginally, parallel execution is still slower than sequential :( Although since atomic primitives are faster than RwLocks in all circumstances, I'll refactor the RwLocks to atomic primitives either ways!